keps/sig-cluster-lifecycle/kubeadm/20190506-kubeadm-machine-output.md

```diff
@@ -0,0 +1,360 @@
+
+Without structured output, even the most seeminlgy innocuous changes could break Terraform, Cluster API, or other software that relies on the results of kubeadm.
+
+### Goals
+
+* Kubeadm should support structured output, including, but not limited to, the following output types:
+  * Buffered
+    * JSON
+    * YAML
+    * Go template
+  * Unbuffered
+    * Text (the current behavior)
+  * Please see [*Buffered vs Unbuffered*](#buffered-vs-unbuffered) for details on what both terms mean in the context of this KEP
+* Kubeadm should support the above output types for commands that include, but not necessarily limited to, the following:
+  * `alpha certs`
+  * `config images list`
+  * `init`
+  * `token list`
+  * `upgrade plan`
+  * `version`
+
+### Non-Goals
+
+* Explicit support for an unbuffered, *short* variant of *Text* output is not necessary at this time as the only command that currently supports this output type is `version`
+* [Parity with kubectl's `-o|--output` flag](#parity-with-kubectl-and-versioned-output)
+* [Versioned output](#parity-with-kubectl-and-versioned-output)
+
+## Proposal
+
+### User Stories
+
+The examples in the user stories are predicated on the existence of:
+
+* [A friendly bootstrap token struct](#a-friendly-bootstrap-token-struct)
+* [Additional functions for Go templates](#go-template-functions)
+
+#### Story 1
+
+The deployment of a multi-node Kubernetes cluster is automated by parsing [the JSON output](#kubeadm-init-json-output) of `kubeadm init` for the emitted `kubeadm join` command.
+
+1. Run `kubeadm init -o json | jq -r '"kubeadm join \(.node0) --token \(.token.id) --discovery-token-ca-cert-hash \(.caCrt)"'`
+
+#### Story 2
+
+A script requires a list of token IDs.
+
+1. Run `kubeadm token list -o '{{range .}}{{println .ID}}{{end}}'`
+
+#### Story 3
+
+A script returns a list of token IDs as they are discovered.
+
+1. Run `kubeadm token list --stream -o '{{println .ID}}'`
+
+#### Story 4
+
+A script needs to process the IDs of all the non-expired tokens:
+
+1. Run `kubeadm token list -o json | jq '.[] | select(.expires | fromdate > now) | .id'`
+
+#### Story 5
+
+A script needs to process the IDs of all the non-expired tokens as the tokens are discovered:
+
+1. Run `kubeadm token list --stream -o json | jq 'select(.expires | fromdate > now) | .id'`
+
+### Implementation Details/Notes/Constraints
+
+#### Details
+
+##### A centralized printer system
+The first design detail is related to how commands handle output...they don't. Or rather, they shouldn't. A command should perform an action and return a result, not print that result. Much like the package [`k8s.io/cli-runtime/pkg/printers`](https://github.com/kubernetes/cli-runtime), a package dedicated to printing will be introduced to kubeadm.
```

Comment:
Got there in the end, but immediately wondered why cli-runtime's package wasn't good enough. Consider linking or slightly rewording so you can get an understanding of why not cli-runtime quickly.

Reply:
How about I state it in the non-goals up top so that it's one of the first things you see?

Reply:
Hi @fabianofranz,

Agreed. I put together the KEP in order to discuss it with the kubeadm working group knowing that the topic of versioned output was going to be the key issue around which the design could pivot. I think the KEP goes out of the way already to describe in detail why cli-runtime wasn't used. It states:

> The natural path to such parity would be to use the same mechanism in kubeadm as used by kubectl, the API machinery package `k8s.io/cli-runtime/pkg/printers`

So it wasn't excluded, rather to the contrary, it was explored fully and declared not viable due to the initial thoughts towards versioned output, a topic I knew may change after the first kubeadm WG call.

Reply:
If I got this right latest discussions / @bart0sh POC, we are going to explore usage of cli-runtime whenever possible/convenient, so I consider this point addressed